File: tests/clients/battery.py
"""The Python half of the cross-client conformance battery.

Runs the same scenarios as ``battery.mjs`` against the same peer, so the two
line up into a matrix. Where this disagrees with the Node driver, two
implementations disagree about the gateway's output — which is the only kind of
defect a single-language suite structurally cannot see.

Usage: python tests/clients/battery.py <url> <http|sse>
"""

from __future__ import annotations

import asyncio
import json
import sys

from mcp import ClientSession
from mcp.client.sse import sse_client
from mcp.client.streamable_http import streamable_http_client

SEPARATORS = "before\u2028middle\u2029after"
UNICODE = "a👩‍👩‍👧‍👦b عربى c éèê d 漢字 e 𝔘𝔫𝔦 f"
PLAIN = "plain-result"
import os
BIG_LENGTH = int(os.environ.get("BIG_LENGTH", 1024 * 1024))

WANT_TOOLS = [
    "echo", "large", "plain", "separators", "shadowedError",
    "slow", "throws", "toolError", "unicode",
]

results: list[dict] = []


def record(name: str, ok: bool, detail: str = "") -> None:
    results.append({"name": name, "ok": bool(ok), "detail": detail})


def text_of(res) -> str | None:
    for c in getattr(res, "content", []) or []:
        if getattr(c, "type", None) == "text":
            return c.text
    return None


async def main() -> None:
    url, kind = sys.argv[1], sys.argv[2]
    ctx = sse_client(url) if kind == "sse" else streamable_http_client(url)
    try:
        async with ctx as streams:
            read, write = streams[0], streams[1]
            async with ClientSession(read, write) as session:
                init = await session.initialize()
                info = getattr(init, "server_info", None) or getattr(init, "serverInfo", None)
                record("connect", True, getattr(info, "name", ""))

                try:
                    names = sorted(t.name for t in (await session.list_tools()).tools)
                    record("tools/list", names == WANT_TOOLS, ",".join(names))
                except Exception as e:
                    record("tools/list", False, f"{type(e).__name__}: {e}"[:60])

                for label, tool, expect in (
                    ("plain text", "plain", PLAIN),
                    ("line separators", "separators", SEPARATORS),
                    ("unicode", "unicode", UNICODE),
                ):
                    try:
                        got = text_of(await session.call_tool(tool, {}))
                        record(label, got == expect, "" if got == expect else repr(got)[:60])
                    except Exception as e:
                        record(label, False, f"{type(e).__name__}: {e}"[:60])

                try:
                    got = text_of(await session.call_tool("large", {}))
                    record("1 MB result", len(got or "") == BIG_LENGTH, f"length {len(got or '')}")
                except Exception as e:
                    record("1 MB result", False, f"{type(e).__name__}: {e}"[:60])

                try:
                    res = await session.call_tool("toolError", {})
                    record("tool error stays a result", getattr(res, "is_error", getattr(res, "isError", None)) is True,
                           f"isError={getattr(res, 'isError', None)}")
                except Exception as e:
                    record("tool error stays a result", False, f"threw {type(e).__name__}"[:50])

                try:
                    res = await session.call_tool("throws", {})
                    record("throwing tool reported",
                           getattr(res, "is_error", getattr(res, "isError", None)) is True or bool(text_of(res)),
                           f"isError={getattr(res, 'isError', None)}")
                except Exception as e:
                    record("throwing tool reported", True, f"rejected {type(e).__name__}"[:40])

                try:
                    args = {
                        "text": UNICODE,
                        "number": -0.5,
                        "flag": False,
                        "nested": {"a": [1, None, True], "b": {"c": SEPARATORS}},
                    }
                    got = json.loads(text_of(await session.call_tool("echo", args)) or "{}")
                    record("arguments round trip", got == args, json.dumps(got, ensure_ascii=False)[:70])
                except Exception as e:
                    record("arguments round trip", False, f"{type(e).__name__}: {e}"[:60])

                try:
                    loop = asyncio.get_running_loop()
                    started = loop.time()
                    got = text_of(await session.call_tool("slow", {}))
                    record("slow call completes", got == "slow-done",
                           f"{int((loop.time() - started) * 1000)}ms")
                except Exception as e:
                    record("slow call completes", False, f"{type(e).__name__}: {e}"[:60])

                try:
                    res = await session.call_tool("shadowedError", {})
                    record("result with an error field", text_of(res) == "ok", repr(text_of(res))[:50])
                except Exception as e:
                    record("result with an error field", False, f"{type(e).__name__}: {e}"[:60])

                try:
                    got = await asyncio.gather(
                        *[session.call_tool("plain", {}) for _ in range(5)]
                    )
                    record("5 concurrent calls", all(text_of(r) == PLAIN for r in got), f"{len(got)} replies")
                except Exception as e:
                    record("5 concurrent calls", False, f"{type(e).__name__}: {e}"[:60])

                try:
                    res = await session.call_tool("noSuchTool", {})
                    record("unknown tool errors", bool(getattr(res, "is_error", getattr(res, "isError", None))),
                           f"isError={getattr(res, 'isError', None)}")
                except Exception as e:
                    record("unknown tool errors", True, f"rejected {type(e).__name__}"[:40])
    except Exception as e:
        detail = f"{type(e).__name__}: {e}"
        subs = getattr(e, "exceptions", None) or []
        while subs:
            inner = subs[0]
            detail = f"{type(inner).__name__}: {inner}"
            subs = getattr(inner, "exceptions", None) or []
        record("connect", False, detail[:140])

    print(json.dumps(results, ensure_ascii=False))


if __name__ == "__main__":
    asyncio.run(main())

File: tests/clients/battery.rb
# The Ruby half of the cross-client conformance battery.
#
# A fifth implementation, and a fifth language runtime. `ruby-mcp-client` is a
# community client rather than the official gem, which is worth knowing: it is
# the kind of client a user might actually reach for, and it makes its own
# choices about parsing.
require 'json'
require 'mcp_client'

SEPARATORS = "before middle after"
UNICODE = "a\u{1F469}‍\u{1F469}‍\u{1F467}‍\u{1F466}b عربى c éèê d 漢字 e \u{1D518}\u{1D52B}\u{1D526} f"
PLAIN = 'plain-result'
BIG_LENGTH = Integer(ENV.fetch('BIG_LENGTH', 1024 * 1024))

$results = []

def record(name, ok, detail = '')
  $results << { name: name, ok: !!ok, detail: detail.to_s[0, 90] }
end

def text_of(res)
  items = res.is_a?(Hash) ? (res['content'] || res[:content]) : nil
  return nil unless items
  item = items.find { |c| (c['type'] || c[:type]) == 'text' }
  item && (item['text'] || item[:text])
end

url = ARGV[0]
kind = ARGV[1]

begin
  config =
    if kind == 'sse'
      MCPClient.sse_config(base_url: url)
    else
      MCPClient.streamable_http_config(base_url: url)
    end
  client = MCPClient.create_client(mcp_server_configs: [config])
  tools = client.list_tools
  record('connect', true, '')
  names = tools.map(&:name).sort
  want = %w[echo large plain separators shadowedError slow throws toolError unicode]
  record('tools/list', names == want, names.join(','))
rescue StandardError => e
  record('connect', false, "#{e.class}: #{e.message}")
  puts JSON.generate($results)
  exit 0
end

def call(client, name, args = {})
  client.call_tool(name, args)
end

[['plain text', 'plain', PLAIN],
 ['line separators', 'separators', SEPARATORS],
 ['unicode', 'unicode', UNICODE]].each do |label, tool, expect|
  begin
    got = text_of(call(client, tool))
    record(label, got == expect, got == expect ? '' : got.inspect)
  rescue StandardError => e
    record(label, false, "#{e.class}: #{e.message}")
  end
end

begin
  got = text_of(call(client, 'large'))
  record('1 MB result', got && got.bytesize == BIG_LENGTH, "length #{got ? got.bytesize : 'nil'}")
rescue StandardError => e
  record('1 MB result', false, "#{e.class}: #{e.message}")
end

begin
  res = call(client, 'toolError')
  flag = res.is_a?(Hash) ? (res['isError'] || res[:isError]) : nil
  record('tool error stays a result', flag == true, "isError=#{flag.inspect}")
rescue StandardError => e
  record('tool error stays a result', false, "threw #{e.class}")
end

begin
  res = call(client, 'throws')
  flag = res.is_a?(Hash) ? (res['isError'] || res[:isError]) : nil
  record('throwing tool reported', flag == true || !text_of(res).nil?, "isError=#{flag.inspect}")
rescue StandardError => e
  record('throwing tool reported', true, "rejected #{e.class}")
end

begin
  args = { 'text' => UNICODE, 'number' => -0.5, 'flag' => false,
           'nested' => { 'a' => [1, nil, true], 'b' => { 'c' => SEPARATORS } } }
  got = JSON.parse(text_of(call(client, 'echo', args)) || '{}')
  record('arguments round trip', got == args, JSON.generate(got)[0, 70])
rescue StandardError => e
  record('arguments round trip', false, "#{e.class}: #{e.message}")
end

begin
  started = Time.now
  got = text_of(call(client, 'slow'))
  record('slow call completes', got == 'slow-done', "#{((Time.now - started) * 1000).to_i}ms")
rescue StandardError => e
  record('slow call completes', false, "#{e.class}: #{e.message}")
end

begin
  res = call(client, 'shadowedError')
  record('result with an error field', text_of(res) == 'ok', text_of(res).inspect)
rescue StandardError => e
  record('result with an error field', false, "#{e.class}: #{e.message}")
end

begin
  threads = 5.times.map { Thread.new { text_of(call(client, 'plain')) } }
  got = threads.map(&:value)
  record('5 concurrent calls', got.all? { |t| t == PLAIN }, "#{got.length} replies")
rescue StandardError => e
  record('5 concurrent calls', false, "#{e.class}: #{e.message}")
end

begin
  res = call(client, 'noSuchTool')
  flag = res.is_a?(Hash) ? (res['isError'] || res[:isError]) : nil
  record('unknown tool errors', !!flag, "isError=#{flag.inspect}")
rescue StandardError => e
  record('unknown tool errors', true, "rejected #{e.class}")
end

puts JSON.generate($results)
